packages/nextjs/mocks/msw/groq/create-dataset.ts

```diff
@@ -0,0 +1,110 @@
+export type Dataset = Array<DatasetItem>;
+
+export type DatasetItem = {
+  _type?: string;
+  _id?: string;
+  [rest: string]: unknown;
+};
+
+type AnyObject = Record<string, unknown>;
+
+export type DatasetOptions = {
+  /**
+   * This list overrides what gets treated as a document.
+   * Documents are always referenced using reference types, instead of being inline.
+   */
+  documentTypes?: string[];
+  /**
+   * This list overrides what gets treated as an inline type.
+   */
+  inlineTypes?: string[];
+};
+
+/**
+ * Builds a flattened dataset from any type of deeply-nested data.
+ */
+export function createDataset(data: Array<AnyObject>, options: DatasetOptions = {}) {
+  const builder = new DatasetBuilder(options);
+  data.forEach((object) => builder.addData(object));
+  return builder.dataset;
+}
+
+class DatasetBuilder {
+  constructor(private options: DatasetOptions) {}
+
+  readonly dataset = [] as Dataset;
+
+  readonly references = new Map<object, DatasetItem>();
```

Comment:
Suppose there's any benefit to using a `WeakMap` here? If this acts something like a cache/temporary store, might be worthwhile (or maybe the GC affordances of a WeakMap don't really provide anything useful here, idk).

Reply:
This class is a "throwaway" class; the `createDataset` method initializes it, uses it, and throws it away. So the references are discarded as soon as the results are calculated.  
I _think_ that `Map` is minutely faster than `WeakMap` so it's probably the best choice for this purpose.